---
title: "An introduction of CancerSubtypes"
author: "Taosheng Xu,Thuc Duy Le"
date: '`r Sys.Date()`'
output:
  html_document:
    number_sections: yes
  pdf_document:
    number_sections: yes
vignette: "%\\VignetteIndexEntry{CancerSubtypes}\n%\\VignetteEncoding{UTF-8}\n%\\VignetteEngine{knitr::rmarkdown}
  \n"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Introduction
With the accumulating wealth of data from large-scale cancer genome project (e.g. TCGA), identifying cancer molecular subtypes is becoming an important topic for cancer genomic data analysis. Distinguishing molecular subtypes for a given cancer greatly assist cancer therapy by moving away from a "one-size-fits-all" approach to medicine.

We devote to generate a customized package for cancer subtype analysis that includes the overall process from dataset processing to result validation with the unified data interface. In CancerSubtypes package, we integrated the typical methods in a framework including genomic data pre-processing, cancer subtypes identification, results validation, visualization and comparison. We implement these clustering methods in an unified input and output interface. So the process of cancer subtypes analysis can be easily conduct in a standard workflow. CancerSubtypes provides the most useful feature selection methods and subtypes validation method and helps users to focus on their cancer genomic data and the results from different methods can be compared and evaluation in visualization way easily.

# Overview of the Package

The features of CancerSubtypes package are listed in below:

- 5 built-in algorithms for cancer subtypes identification.
- 4 built-in feature selection methods for genomic dataset.
- 4 built-in methods for result validation and visualization.
- Unified input and output interface to perform all clustering methods.
- Provides a framework to identify cancer subtypes and result validation.
- Visualization results to help in the interpretation of the biological meaning of identified cancer subtypes.


# Data processing
For the basic data processing, CancerSubtypes provides the methods for data distribution check, imputation and normalization. In most cases, feature dimension reduction is needed and necessary for most high dimension genomic datasets to remove irrelevant features and reduce noise. There are four feature selection methods (Variance-Var, Median Absolute Deviation-MAD, COX model, Principal Component Analysis-PCA) in CancerSubtypes package. Feature selection by Var or MAD are the common way adopted in many cancer genomic studies. However, selecting genomic features by survival analysis Cox model is a novel approach that can gain the important survival related features. Feature selection by PCA is to discard some tail PCs in principal component space. All the data processing methods possess the same input and output data format.

## Basic processing

### Analysis the raw data by check the data distribution.
```{r, message = FALSE}
library(CancerSubtypes)
data(GeneExp)
data.checkDistribution(GeneExp)
```

### Data imputation for features with missing values (NAs)
```{r}
Data=matrix(runif(1000),nrow = 50,ncol = 20) ###Generate an random matrix
geneName=paste("Gene", 1:50, sep = " ")
sampleName=paste("Sample", 1:20, sep = " ")
rownames(Data)=geneName
colnames(Data)=sampleName 
index=sample(c(1:1000),60)
Data[index]=NA
which(is.na(Data))
res1=data.imputation(Data,fun="median")
which(is.na(res1))
res2=data.imputation(Data,fun="mean")
which(is.na(res2))
res3=data.imputation(Data,fun="microarray")
which(is.na(res3))
```
### Data normalization.
```{r}
data(GeneExp)
result1=data.normalization(GeneExp,type="feature_Median",log2=FALSE)
result2=data.normalization(GeneExp,type="feature_Mean",log2=FALSE)
result3=data.normalization(GeneExp,type="feature_zsocre",log2=FALSE)
result4=data.normalization(GeneExp,type="sample_zsocre",log2=FALSE)
```
## Feature selection

### Biological feature selection based on the most variance.

```{r}
data(GeneExp)
data(miRNAExp)
data1=FSbyVar(GeneExp, cut.type="topk",value=1000)###The top 1000 most variance features will be selected.
data2=FSbyVar(GeneExp, cut.type="cutoff",value=0.5)###The features with (variance>0.5) are selected.
data3=FSbyVar(miRNAExp, cut.type="topk",value=300)###The top 300 most variance features will be selected.
data4=FSbyVar(miRNAExp, cut.type="cutoff",value=0.5)###The features with (variance>0.5) are selected.
```

### Biological feature selection based on the most variant Median Absolute Deviation (MAD).
```{r}
data(GeneExp)
data(miRNAExp)
data1=FSbyMAD(GeneExp, cut.type="topk",value=1000)
data2=FSbyMAD(GeneExp, cut.type="cutoff",value=0.5)
data3=FSbyMAD(miRNAExp, cut.type="topk",value=300)
data4=FSbyMAD(miRNAExp, cut.type="cutoff",value=0.5)
```

### Biological feature selection based on Cox regression model.
```{r}
data(GeneExp)
data(miRNAExp)
data(time)
data(status)
data1=FSbyCox(GeneExp,time,status,cutoff=0.05)
data2=FSbyCox(miRNAExp,time,status,cutoff=0.05)
```
### Biological feature dimension reduction and extraction based on Principal Component Analysis.

```{r}
data(GeneExp)
data(miRNAExp)
data1=FSbyPCA(GeneExp, PC_percent=0.9,scale = TRUE)
data2=FSbyPCA(miRNAExp, PC_percent=0.9,scale = TRUE)
```
# Clustering methods for Cancer subtypes identification.

CancerSubtypes integrates four common and high-cited methods (CC, CNMF, iCluster, SNF) for caner subtypes identification from single-genomic to multi-genomics data. The methods of CC and CNMF are designed for single-genomic dataset (always gene expression data). The iCluster and SNF focus on the multiple genomic data analysis in the context of the rapid development of genome sequencing. All these methods are rewritten with unified input and output interfaces for the purpose of embedding the cancer subtype identification in a workflow with the convenient upstream and downstream analysis.

## Execute Consensus Clustering for cancer subtype identification

Consensus clustering (CC, 2003) as an unsupervised subtypes discovery method, was a frequently used and valuable approach in many genomic studies and have lots of successful application.

```{r}
### The input dataset is single gene expression matrix.
data(GeneExp)
result=ExecuteCC(clusterNum=3,d=GeneExp,maxK=10,clusterAlg="hc",distance="pearson",title="GBM")

### The input dataset is multi-genomics data as a list
data(GeneExp)
data(miRNAExp)
GBM=list(GeneExp=GeneExp,miRNAExp=miRNAExp)
result=ExecuteCC(clusterNum=3,d=GBM,maxK=10,clusterAlg="hc",distance="pearson",title="GBM")
```

## Execute consensus Non-negative matrix factorization for cancer subtype identification

Non-negative matrix factorization (CNMF, 2004), as an effective dimension reduction method, was used in distinguishing molecular patterns for high-dimensional genomic data and provided a powerful method for class discovery.

```{r}
### The input dataset is single gene expression matrix.
data(GeneExp)
result=ExecuteCNMF(GeneExp,clusterNum=3,nrun=30)

### The input dataset is multi-genomics data as a list
data(GeneExp)
data(miRNAExp)
GBM=list(GeneExp=GeneExp,miRNAExp=miRNAExp)
result=ExecuteCNMF(GBM,clusterNum=3,nrun=30)
```

## Execute Integrative clustering for cancer subtype identification

Integrative clustering (iCluster, 2009) used a joint latent variable model for integrative clustering for multiple types of omics data.

```{r,fig.show ='hide'}
data(GeneExp)
data(miRNAExp)
data1=FSbyVar(GeneExp, cut.type="topk",value=1000)
data2=FSbyVar(miRNAExp, cut.type="topk",value=300)
```
```{r}
GBM=list(GeneExp=data1,miRNAExp=data2)
result=ExecuteiCluster(datasets=GBM, k=3, lambda=list(0.44,0.33,0.28))
```


## Execute Similarity network fusion for cancer subtype identification

Similarity network fusion (SNF, 2014) is a computational method on fusion similarity network for aggregating multi-omics data.

```{r}
data(GeneExp)
data(miRNAExp)
GBM=list(GeneExp=GeneExp,miRNAExp=miRNAExp)
result=ExecuteSNF(GBM, clusterNum=3, K=20, alpha=0.5, t=20)
```

## Execute  SNF and CC for cancer subtype identification

We propose to combine the SNF and CC together to generate a new cancer subtypes identification method.

```{r}
data(GeneExp)
data(miRNAExp)
data1=FSbyCox(GeneExp,time,status,cutoff=0.05)
data2=FSbyCox(miRNAExp,time,status,cutoff=0.05)
GBM=list(GeneExp=data1,miRNAExp=data2)
result=ExecuteSNF.CC(GBM, clusterNum=3, K=20, alpha=0.5, t=20,
                     maxK = 10, pItem = 0.8,reps=500, 
                     title = "GBM3", plot = "png", 
                     finalLinkage ="average")
```

# Results validation, interpretation and visualization for the identified cancer subtypes.

The identified cancer subtypes by the computational methods should be in accordance with biological meanings and reveal the distinct molecular patterns. CancerSubtypes provides three statistic methods (survival analysis, Statistical significance of clustering, Different expression) and a computational method (Silhouette width) that are adopted in many cancer genomic studies for the identified cancer subtypes evaluation and visualization.

## Survival analysis

Survival analysis is used to judge the different survival patterns between subtypes.

```{r}
data(GeneExp)
data(miRNAExp)
data(time)
data(status)
data1=FSbyCox(GeneExp,time,status,cutoff=0.05)
data2=FSbyCox(miRNAExp,time,status,cutoff=0.05)
GBM=list(GeneExp=data1,miRNAExp=data2)

##### 1.ExecuteCC
result1=ExecuteCC(clusterNum=3,d=GBM,maxK=10,clusterAlg="hc",
                   distance="pearson",title="GBM1")
group1=result1$group
distanceMatrix1=result1$distanceMatrix
p_value=survAnalysis(mainTitle="GBM1",time,status,group1,
                     distanceMatrix1,similarity=TRUE)

#### 2.ExecuteSNF
result2=ExecuteSNF(GBM, clusterNum=3, K=20, alpha=0.5, t=20)
group2=result2$group
distanceMatrix2=result2$distanceMatrix
p_value=survAnalysis(mainTitle="GBM2",time,status,group2,
                     distanceMatrix2,similarity=TRUE)
                     
#### 3.ExecuteSNF.CC
result3=ExecuteSNF.CC(GBM, clusterNum=3, K=20, alpha=0.5, t=20,
                     maxK = 10, pItem = 0.8,reps=500, 
                     title = "GBM3", plot = "png", 
                     finalLinkage ="average")
group3=result3$group
distanceMatrix3=result3$distanceMatrix
p_value=survAnalysis(mainTitle="GBM3",time,status,group3,
                     distanceMatrix3,similarity=TRUE)

```

## Statistical significance of clustering

Statistical significance of clustering is a pure statistical approach to test the significance difference data distribution between subtypes. Different expression  is to test the expression difference between each subtypes and an reference group (always a set of normal samples).

```{r,warning=FALSE}
data(GeneExp)
data(miRNAExp)
data(time)
data(status)
GBM=list(GeneExp=GeneExp,miRNAExp=miRNAExp)
result=ExecuteSNF(GBM, clusterNum=3, K=20, alpha=0.5, t=20,plot = FALSE)
group=result$group
sigclust1=sigclustTest(GeneExp,group, nsim=1000, nrep=1, icovest=1)
sigclust1
sigclust2=sigclustTest(GeneExp,group, nsim=1000, nrep=1, icovest=3)
sigclust2
sigclust3=sigclustTest(miRNAExp,group, nsim=1000, nrep=1, icovest=1)
sigclust3
```

##Different expression 

Different expression is to test the expression difference between each subtypes and an reference group (always a set of normal samples).

```{r}
data(GeneExp)
data(miRNAExp)
GBM=list(GeneExp=GeneExp,miRNAExp=miRNAExp)
result=ExecuteSNF(GBM, clusterNum=3, K=20, alpha=0.5, t=20,plot = FALSE)
group=result$group
######Fabricate a normal group by extracting some samples from the cancer dataset 
######for demonstrating the examples.
Normal_Data=GeneExp[,sample(1:100,20)]
result=DiffExp.limma(Tumor_Data=GeneExp,Normal_Data=Normal_Data,group=group,topk=NULL,RNAseq=FALSE)

```

## Silhouette width
Silhouette width is used to measure how similar a sample is matched to its identified subtype compared to other subtypes, a high value indicates that the sample is well matched.

```{r}
data(GeneExp)
data(miRNAExp)
GBM=list(GeneExp=GeneExp,miRNAExp=miRNAExp)
result=ExecuteSNF(GBM, clusterNum=3, K=20, alpha=0.5, t=20,plot = FALSE)
###Similarity smapel mateix
sil=silhouette2(result$group, result$distanceMatrix)
plot(sil)
###If use the silhouette() in R package "cluter", the result 
###is meaningless because this is a similarity matrix.
sil1=silhouette(result$group, result$distanceMatrix)
plot(sil1)  ##wrong result
```

# Conclusions

The Bioconductor CancerSubtypes R package provides a suite of cancer subtypes analysis tools and embeds the analysis in a standardized framework. This suite tools including data processing, clustering methods and result validation and visualization can meet most kinds of users (e.g. Biologists, Bioinformaticians and Computer scientist) for cancer subtype analysis from cancer genomic data analysis to a new cancer subtype identification method assessment. It provides a powerful way to analyze cancer subtype on genome-wide scale.

# References


